### PR TITLE
Don't run neutron api service as root

### PR DESCRIPTION
**Pull request by booxter**

### CODE REVIEW

**Comment by abays** [comment]
> @booxter: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> ci/prow/neutron-operator-build-deploy-kuttl 	[3631589](https://github.com/openstack-k8s-operators/neutron-operator/commit/363158921805564c00ec9cad6806c0fe362c2a4e) 	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_neutron-operator/163/pull-ci-openstack-k8s-operators-neutron-operator-main-neutron-operator-build-deploy-kuttl/1671447380999802880) 	false 	`/test neutron-operator-build-deploy-kuttl`
> 
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=openstack-k8s-operators&repo=neutron-operator&pr=163). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:booxter).

I'm not sure what is going on with CI here:

```
{  error occurred handling build src: build not successful after 5 attempts: [the build src failed after 37s with reason FetchImageContentFailed: Failed to extract image content., the build src failed after 38s with reason FetchImageContentFailed: Failed to extract image content.]}
```

**Comment by abays** [comment]
> > @booxter: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> > Test name 	Commit 	Details 	Required 	Rerun command
> > ci/prow/neutron-operator-build-deploy-kuttl 	[3631589](https://github.com/openstack-k8s-operators/neutron-operator/commit/363158921805564c00ec9cad6806c0fe362c2a4e) 	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_neutron-operator/163/pull-ci-openstack-k8s-operators-neutron-operator-main-neutron-operator-build-deploy-kuttl/1671447380999802880) 	false 	`/test neutron-operator-build-deploy-kuttl`
> > [Full PR test history](https://prow.ci.openshift.org/pr-history?org=openstack-k8s-operators&repo=neutron-operator&pr=163). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:booxter).
> 
> I'm not sure what is going on with CI here:
> 
> ```
> {  error occurred handling build src: build not successful after 5 attempts: [the build src failed after 37s with reason FetchImageContentFailed: Failed to extract image content., the build src failed after 38s with reason FetchImageContentFailed: Failed to extract image content.]}
> ```

Oh, this has cleared up now.

**Comment by abays** [comment]
/test neutron-operator-build-deploy-kuttl

> @booxter: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> Test name 	Commit 	Details 	Required 	Rerun command
> ci/prow/neutron-operator-build-deploy-kuttl 	[3631589](https://github.com/openstack-k8s-operators/neutron-operator/commit/363158921805564c00ec9cad6806c0fe362c2a4e) 	[link](https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openstack-k8s-operators_neutron-operator/163/pull-ci-openstack-k8s-operators-neutron-operator-main-neutron-operator-build-deploy-kuttl/1671495855045087232) 	false 	`/test neutron-operator-build-deploy-kuttl`
> 
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=openstack-k8s-operators&repo=neutron-operator&pr=163). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:booxter).

KUTTL test is failing like so:

```
case.go:366: resource Pod:openstack/: .metadata.labels.service: key is missing from map
```

File: controllers/neutronapi_controller.go [hunk 1]
```diff
@@ -528,7 +528,7 @@ func (r *NeutronAPIReconciler) reconcileNormal(ctx context.Context, instance *ne
 	rbacRules := []rbacv1.PolicyRule{
 		{
 			APIGroups:     []string{"security.openshift.io"},
-			ResourceNames: []string{"anyuid", "privileged"},
+			ResourceNames: []string{"anyuid"},
```
Comment (karelyatin) [inline controllers/neutronapi_controller.go:531]: only once concern how we deal with extra mounts that allows host path https://github.com/openstack-k8s-operators/neutron-operator/blob/efd4186556fec42b73d22cfd13ae4dac8bb90488/api/bases/neutron.openstack.org_neutronapis.yaml#L1179-L1188 ?

File: pkg/neutronapi/volumes.go [hunk 2]
```diff
@@ -39,25 +39,9 @@ func GetInitVolumeMounts(extraVol []neutronv1beta1.NeutronExtraVolMounts, svc []
 //	mechanism.
 func GetAPIVolumes(name string, extraVol []neutronv1beta1.NeutronExtraVolMounts, svc []storage.PropagationType) []corev1.Volume {
 	var scriptsVolumeDefaultMode int32 = 0755
-	var config0640AccessMode int32 = 0640
+	var config0644AccessMode int32 = 0644
 
 	res := []corev1.Volume{
-		{
-			Name: "etc-machine-id",
```
Comment (karelyatin) [inline pkg/neutronapi/volumes.go:46]: These being dropped as part of https://github.com/openstack-k8s-operators/neutron-operator/pull/167

File: pkg/neutronapi/volumes.go [hunk 3]
```diff
@@ -99,16 +83,6 @@ func GetAPIVolumes(name string, extraVol []neutronv1beta1.NeutronExtraVolMounts,
 // GetAPIVolumeMounts - Neutron API VolumeMounts
 func GetAPIVolumeMounts(extraVol []neutronv1beta1.NeutronExtraVolMounts, svc []storage.PropagationType) []corev1.VolumeMount {
 	res := []corev1.VolumeMount{
-		{
-			Name:      "etc-machine-id",
-			MountPath: "/etc/machine-id",
-			ReadOnly:  true,
```
Comment (karelyatin) [inline pkg/neutronapi/volumes.go:105]: same

File: controllers/neutronapi_controller.go (hunk 1 above)
Comment (abays) [reply controllers/neutronapi_controller.go:531]: One option is to use this security context constraint instead:

```
hostmount-anyuid | Provides all the features of the restricted SCC, but allows host mounts and running as any UID and any GID on the system.
```

https://docs.openshift.com/container-platform/4.12/authentication/managing-security-context-constraints.html#default-sccs_configuring-internal-oauth

**Comment by booxter** [comment]
FYI this is not ready, I need to re-test it properly, I think there are some remnants of runAsUser=0 that can be cleaned up before merge.

File: controllers/neutronapi_controller.go (hunk 1 above)
Comment (booxter) [reply controllers/neutronapi_controller.go]: @karelyatin (reviving a very old PR) I think that maybe it's a suggestion that we should rethink how we extend neutronapi service, instead of leaving hostmount enabled? Would e.g. switching to a PV with custom driver code be a better path here? Let's discuss it some time.

**Comment by booxter** [comment]
@karelyatin I posted a new version of the draft to remove privileges from neutron pods.

It is not ready to merge (I expect all jobs to fail - functional included) but it seems to do the right thing with `make run-with-webhook` (all pods are successfully started and running). Obviously, I will need to fix the test failures as well as polish the code (lots of TODOs there).

But I would love to see your initial feedback on the approach here first, since you are more knowledgeable about kolla and related topics than me. I'd like to understand if the path is generally correct before spending time polishing the draft.

tl;dr the PR did the following:
- remove all kolla stuff (no more config copying / merging)
- the full config with all secrets is now generated into ConfigMap prior to starting helper and main neutron pods (should be switched to Secret)
- the configs are now mounted into /neutron.conf.d similar to edpm-ansible ovn roles (ovn, sriov)
- SCC fields are added to pods to use neutron:neutron to start processes - this means we need anyuid SCC

We may later also discuss if / why we need hostmount-anyuid (I know it's for ExtraMounts but maybe we should have 3rd parties mount their customizations via PVCs?). This is a separate and more broad topic, so leaving this out.

Thanks!

**Comment by booxter** [comment]
@karelyatin regarding the need to run debug commands as root, you can always do

```
oc debug pod/... --as-root
```

to get a root shell.

---

Thanks for the initial feedback, I will now split the PR into pieces (one for kolla cleanup, one for privileges, maybe more pieces).

**Comment by booxter** [comment]
kolla removal split here https://github.com/openstack-k8s-operators/neutron-operator/pull/221

**Comment by booxter** [comment]
de-escalation piece of the PR: https://github.com/openstack-k8s-operators/neutron-operator/pull/223

This PR is now redundant. Closing.
